ReactWindows/ReactNative.Shared/Common/AccessibilityHelper.cs

```diff
@@ -0,0 +1,229 @@
+        /// <param name="accessibilityView"></param>
+        private static void SetChildrenAccessibilityView(AutomationPeer parentAutomationPeer, AccessibilityView accessibilityView)
+        {
+            if (parentAutomationPeer?.GetChildren() == null)
+            {
+                return;
+            }
+
+            foreach (AutomationPeer childAutomationPeer in parentAutomationPeer.GetChildren())
+            {
+                if (childAutomationPeer is FrameworkElementAutomationPeer childFEAutomationPeer)
+                {
+                    if (childFEAutomationPeer?.Owner != null)
+                    {
+                        AutomationProperties.SetAccessibilityView(childFEAutomationPeer.Owner, accessibilityView);
+                    }
+                }
+                else if (childAutomationPeer is ItemAutomationPeer childIAutomationPeer)
+                {
+                    if (childIAutomationPeer?.Item is UIElement childUIElement)
+                    {
+                        AutomationProperties.SetAccessibilityView(childUIElement, accessibilityView);
+                    }
+                }
+
+                SetChildrenAccessibilityView(childAutomationPeer, accessibilityView);
+            }
+        }
+
+        /// <summary>
+        /// Returns true if in uiElementAutomationPeer parents chain there is an UIElement with
+        /// ImportantForAccessibility value of Yes or NoHideDescendants.
+        /// </summary>
+        /// <param name="uiElementAutomationPeer"></param>
+        /// <returns></returns>
+        private static bool IsHiddenElement(AutomationPeer uiElementAutomationPeer)
+        {
+            if (uiElementAutomationPeer == null)
+            {
+                return false;
+            }
+
+            ImportantForAccessibility importantForAccessibilityAttached = ImportantForAccessibility.Auto;
+            var parentAutomationPeer = uiElementAutomationPeer.Navigate(AutomationNavigationDirection.Parent) as AutomationPeer;
+            if (parentAutomationPeer is FrameworkElementAutomationPeer parentFEAutomationPeer)
+            {
+                if (parentFEAutomationPeer?.Owner != null)
+                {
+                    importantForAccessibilityAttached = GetImportantForAccessibilityAttached(parentFEAutomationPeer.Owner);
+                }
+            }
+            else if (parentAutomationPeer is ItemAutomationPeer parentIAutomationPeer)
+            {
+                if (parentIAutomationPeer?.Item is UIElement parentUIElement)
+                {
+                    importantForAccessibilityAttached = GetImportantForAccessibilityAttached(parentUIElement);
+                }
+            }
+
+            if (importantForAccessibilityAttached == ImportantForAccessibility.Yes ||
+                importantForAccessibilityAttached == ImportantForAccessibility.NoHideDescendants)
+            {
+                return true;
+            }
+            else
```

Comment:
>else [](start = 12, length = 4)

this else merges three cases: when the parent is null, when the parent is an unsupported, unexpected class which is not even AutomationPeer-based, when the parent is AutomationPeer. Maybe it would make code more maintainable and also we could insert logging or error raising in case of non-AutomationPeer if we split those three cases?

Reply:
I reworked the code using switch with pattern matching so that all the cases are outlined clearly. Should make it easier to maintain.

---
In reply to: [163113056](https://github.com/Microsoft/react-native-windows/pull/1595#discussion_r163113056) [](ancestors = 163113056)